### PR TITLE
Lint

### PR DESCRIPTION
**Pull request by fabiosantoscode**
Porting the 4c5b9d3 commit, little by little.

I guessed that a lint must have been used to produce the changeset, and reverse-engineered an eslint config.

TODO:

- [ ] add an npm script to run the lint (with autofix?)
- [ ] use the [husky](npmjs.com/husky) package to run the lint automatically as a precommit hook
- [ ] <s>investigate performance issues in mangling htmlminifier.js and ember.prod.js</s>

### CODE REVIEW

File: lib/ast.js [hunk 1]
```diff
@@ -646,11 +644,10 @@ var AST_Definitions = DEFNODE("Definitions", "definitions", {
...
     _walk: function(visitor) {
-        return visitor._visit(this, function(){
-            var definitions = this.definitions;
-            for (var i = 0, len = definitions.length; i < len; i++) {
-                definitions[i]._walk(visitor);
-            }
+        return visitor._visit(this, function() {
+            this.definitions.forEach(function(defn) {
+                defn._walk(visitor);
+            });
```
Comment (kzc) [inline lib/ast.js]: @fabiosantoscode I was hoping these loop code changes wouldn't be ported. They slow down execution in my tests.

File: lib/ast.js [hunk 2]
```diff
@@ -878,11 +875,10 @@ var AST_Array = DEFNODE("Array", "elements", {
...
     _walk: function(visitor) {
-        return visitor._visit(this, function(){
-            var elements = this.elements;
-            for (var i = 0, len = elements.length; i < len; i++) {
-                elements[i]._walk(visitor);
-            }
+        return visitor._visit(this, function() {
+            this.elements.forEach(function(element) {
+                element._walk(visitor);
+            });
```
Comment (kzc) [inline lib/ast.js]: ditto

File: lib/ast.js [hunk 3]
```diff
@@ -893,11 +889,10 @@ var AST_Object = DEFNODE("Object", "properties", {
...
     _walk: function(visitor) {
-        return visitor._visit(this, function(){
-            var properties = this.properties;
-            for (var i = 0, len = properties.length; i < len; i++) {
-                properties[i]._walk(visitor);
-            }
+        return visitor._visit(this, function() {
+            this.properties.forEach(function(prop) {
+                prop._walk(visitor);
+            });
```
Comment (kzc) [inline lib/ast.js]: ditto

File: lib/ast.js [hunk 4]
```diff
@@ -158,11 +158,10 @@ function walk_body(node, visitor) {
...
     if (body instanceof AST_Node) {
         body._walk(visitor);
-    }
-    else for (var i = 0, len = body.length; i < len; i++) {
-        body[i]._walk(visitor);
-    }
-};
+    } else body.forEach(function(node) {
+        node._walk(visitor);
+    });
```
Comment (kzc) [inline lib/ast.js]: I prefer the loop over `forEach` for performance reasons on older JS engines.

File: lib/ast.js [hunk 5]
```diff
@@ -406,22 +405,21 @@ var AST_Lambda = DEFNODE("Lambda", "name argnames uses_arguments is_generator as
     },
     args_as_names: function () {
         var out = [];
-        for (var i = 0; i < this.argnames.length; i++) {
-            if (this.argnames[i] instanceof AST_Destructuring) {
-                out = out.concat(this.argnames[i].all_symbols());
+        this.argnames.forEach(function(argname) {
+            if (argname instanceof AST_Destructuring) {
+                out = out.concat(argname.all_symbols());
```
Comment (kzc) [inline lib/ast.js]: prefer loop to forEach for perf reasons

File: lib/ast.js [hunk 6]
```diff
@@ -406,22 +405,21 @@ var AST_Lambda = DEFNODE("Lambda", "name argnames uses_arguments is_generator as
...
-        return visitor._visit(this, function(){
+        return visitor._visit(this, function() {
             if (this.name) this.name._walk(visitor);
-            var argnames = this.argnames;
-            for (var i = 0, len = argnames.length; i < len; i++) {
-                argnames[i]._walk(visitor);
-            }
+            this.argnames.forEach(function(argname) {
+                argname._walk(visitor);
+            });
```
Comment (kzc) [inline lib/ast.js]: prefer loop to forEach for perf reasons

**Comment by fabiosantoscode** [comment]
My main reason to adopt these changes was to help cherry-picking changes from the main fork. When a sweeping change like this is done, and it touches a lot of places in the codebase, it would make cherry-picks not automatically merge.

My other reason was that the whitespace changes indeed make sense and the removed semicolons are on point.

> I prefer the loop over forEach for performance reasons on older JS engines.

I agree with you, but before I change them back, what JS engines are we talking about? How old are they? And should we care about them?

According to the benchmarks, there are speed improvements in this changeset, except for ember and htmlminifier, where the mangle step jumped up by several seconds.

Benchmarks:

```
>node -v
v8.9.1
```

**before**

```
https://code.jquery.com/jquery-3.2.1.js
- parse: 0.652s
- rename: 0.000s
- compress: 6.909s
- scope: 0.487s
- mangle: 0.720s
- properties: 0.000s
- output: 0.379s
- total: 9.147s

Original: 268039 bytes
Uglified: 86099 bytes
GZipped:  29999 bytes
SHA1 sum: 40de304ba426a65eb542b40462fe6db11ba3c4bf

https://cdnjs.cloudflare.com/ajax/libs/angular.js/1.6.4/angular.js
- parse: 2.100s
- rename: 0.000s
- compress: 11.297s
- scope: 0.586s
- mangle: 0.818s
- properties: 0.000s
- output: 0.398s
- total: 15.199s

Original: 1249863 bytes
Uglified: 174094 bytes
GZipped:  59939 bytes
SHA1 sum: 629f054ce603cca2a2dcf0ce58631bed9d81555c

https://cdnjs.cloudflare.com/ajax/libs/mathjs/3.9.0/math.js
- parse: 4.673s
- rename: 0.000s
- compress: 16.866s
- scope: 0.354s
- mangle: 0.485s
- properties: 0.000s
- output: 0.298s
- total: 22.676s

Original: 1590107 bytes
Uglified: 466990 bytes
GZipped:  118814 bytes
SHA1 sum: 8ce39868025e9404d83cef1b7ebf4aef7252feed

https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.js
- parse: 0.216s
- rename: 0.000s
- compress: 1.803s
- scope: 0.187s
- mangle: 0.402s
- properties: 0.000s
- output: 0.122s
- total: 2.730s

Original: 69707 bytes
Uglified: 36836 bytes
GZipped:  9582 bytes
SHA1 sum: aba42e2bdbcbc6d6b1c35e9aaaa1dca7367d37f6

https://unpkg.com/react@15.3.2/dist/react.js
- parse: 1.813s
- rename: 0.000s
- compress: 10.740s
- scope: 0.570s
- mangle: 0.798s
- properties: 0.000s
- output: 0.459s
- total: 14.380s

Original: 701412 bytes
Uglified: 205595 bytes
GZipped:  62120 bytes
SHA1 sum: 9972b78d2f796246f82bc0952c403f777e7a4f59

http://builds.emberjs.com/tags/v2.11.0/ember.prod.js
- parse: 4.263s
- rename: 0.000s
- compress: 12.858s
- scope: 0.713s
- mangle: 0.786s
- properties: 0.000s
- output: 0.453s
- total: 19.073s

Original: 1852178 bytes
Uglified: 526693 bytes
GZipped:  128766 bytes
SHA1 sum: 7c4262ae71b582892607fcbc09d74f4be3c0d2b8

https://cdn.jsdelivr.net/lodash/4.17.4/lodash.js
- parse: 1.634s
- rename: 0.000s
- compress: 10.418s
- scope: 0.333s
- mangle: 0.619s
- properties: 0.000s
- output: 0.277s
- total: 13.281s

Original: 539590 bytes
Uglified: 70167 bytes
GZipped:  24699 bytes
SHA1 sum: 3aba525fcba83326872456ddf655705b901b6ee1

https://cdnjs.cloudflare.com/ajax/libs/d3/4.5.0/d3.js
- parse: 2.847s
- rename: 0.000s
- compress: 12.618s
- scope: 0.510s
- mangle: 0.715s
- properties: 0.000s
- output: 0.356s
- total: 17.048s

Original: 451131 bytes
Uglified: 212046 bytes
GZipped:  70690 bytes
SHA1 sum: 33ccb8f4cc70e1f85bd67f1a4a24d833f07c3c6d

https://raw.githubusercontent.com/kangax/html-minifier/v3.5.7/dist/htmlminifier.js
- parse: 4.508s
- rename: 0.000s
- compress: 15.723s
- scope: 0.453s
- mangle: 0.639s
- properties: 0.000s
- output: 0.425s
- total: 21.748s

Original: 1063075 bytes
Uglified: 510032 bytes
GZipped:  157422 bytes
SHA1 sum: a4493af20cebe4ed1135a435e47b52ea9756d1f0
```

**after**

```
https://code.jquery.com/jquery-3.2.1.js
- parse: 0.571s
- rename: 0.000s
- compress: 6.426s
- scope: 0.441s
- mangle: 0.692s
- properties: 0.000s
- output: 0.367s
- total: 8.497s

Original: 268039 bytes
Uglified: 86099 bytes
GZipped:  29999 bytes
SHA1 sum: 40de304ba426a65eb542b40462fe6db11ba3c4bf

https://cdnjs.cloudflare.com/ajax/libs/angular.js/1.6.4/angular.js
- parse: 1.803s
- rename: 0.000s
- compress: 10.951s
- scope: 0.540s
- mangle: 1.416s
- properties: 0.000s
- output: 0.468s
- total: 15.178s

Original: 1249863 bytes
Uglified: 174094 bytes
GZipped:  59939 bytes
SHA1 sum: 629f054ce603cca2a2dcf0ce58631bed9d81555c

https://cdnjs.cloudflare.com/ajax/libs/mathjs/3.9.0/math.js
- parse: 4.047s
- rename: 0.000s
- compress: 17.205s
- scope: 0.567s
- mangle: 1.971s
- properties: 0.000s
- output: 0.606s
- total: 24.396s

Original: 1590107 bytes
Uglified: 466990 bytes
GZipped:  118814 bytes
SHA1 sum: 8ce39868025e9404d83cef1b7ebf4aef7252feed

https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.js
- parse: 0.237s
- rename: 0.000s
- compress: 1.561s
- scope: 0.112s
- mangle: 0.311s
- properties: 0.000s
- output: 0.153s
- total: 2.374s

Original: 69707 bytes
Uglified: 36836 bytes
GZipped:  9582 bytes
SHA1 sum: aba42e2bdbcbc6d6b1c35e9aaaa1dca7367d37f6

https://unpkg.com/react@15.3.2/dist/react.js
- parse: 1.730s
- rename: 0.000s
- compress: 10.765s
- scope: 0.462s
- mangle: 0.865s
- properties: 0.000s
- output: 0.469s
- total: 14.291s

Original: 701412 bytes
Uglified: 205595 bytes
GZipped:  62120 bytes
SHA1 sum: 9972b78d2f796246f82bc0952c403f777e7a4f59

http://builds.emberjs.com/tags/v2.11.0/ember.prod.js
- parse: 4.159s
- rename: 0.000s
- compress: 13.650s
- scope: 0.503s
- mangle: 3.075s
- properties: 0.000s
- output: 0.845s
- total: 22.232s

Original: 1852178 bytes
Uglified: 526693 bytes
GZipped:  128766 bytes
SHA1 sum: 7c4262ae71b582892607fcbc09d74f4be3c0d2b8

https://cdn.jsdelivr.net/lodash/4.17.4/lodash.js
- parse: 1.298s
- rename: 0.000s
- compress: 10.133s
- scope: 0.367s
- mangle: 0.637s
- properties: 0.000s
- output: 0.278s
- total: 12.713s

Original: 539590 bytes
Uglified: 70167 bytes
GZipped:  24699 bytes
SHA1 sum: 3aba525fcba83326872456ddf655705b901b6ee1

https://cdnjs.cloudflare.com/ajax/libs/d3/4.5.0/d3.js
- parse: 2.279s
- rename: 0.000s
- compress: 12.817s
- scope: 0.477s
- mangle: 0.802s
- properties: 0.000s
- output: 0.439s
- total: 16.814s

Original: 451131 bytes
Uglified: 212046 bytes
GZipped:  70690 bytes
SHA1 sum: 33ccb8f4cc70e1f85bd67f1a4a24d833f07c3c6d

https://raw.githubusercontent.com/kangax/html-minifier/v3.5.7/dist/htmlminifier.js
- parse: 3.667s
- rename: 0.000s
- compress: 15.868s
- scope: 0.659s
- mangle: 4.596s
- properties: 0.000s
- output: 2.083s
- total: 26.873s

Original: 1063075 bytes
Uglified: 510032 bytes
GZipped:  157422 bytes
SHA1 sum: a4493af20cebe4ed1135a435e47b52ea9756d1f0
```

**Comment by kzc** [comment]
> My other reason was that the whitespace changes indeed make sense and the removed semicolons are on point.

I have no issue with making the formatting of the code match `uglify-js`. It's just the cases where the actual code is changed that gives me pause.

Many of the bug fix cherry-picks I merged were far from automatic. I had to port some of them line by line as the structure of the code changed considerably from `uglify-js` and in some cases had to add additional code and tests for ES6+ functionality. 

And because I chose not to add some ES5 optimizations, around a half dozen bugs never appeared in `terser` and code changes were not necessary - although I did add the tests anyway.

**Comment by fabiosantoscode** [comment]
@kzc you're making a lot of sense - I'm not willing to incur the bug risk.

I've remade the commits. This time I just wrote the eslintConfig and ran eslint --fix.

**Comment by kzc** [comment]
@fabiosantoscode Thanks. I believe that's the safest tradeoff that will ease future upstream bug fix merges somewhat.

BTW, I see `uglify-js` got rid of `mocha` in favor of minimal custom rewrite. I don't think `terser` needs that. `mocha` appears to be working fine.

**Comment by fabiosantoscode** [comment]
Mocha is one of the most stable projects I've come to use in nodejs. I've used it for half a decade and the interface never changed.

Now what I'm struggling with is the husky module. It uses ES6, so node 0.10-0.12 doesn't support it. I'll have to find an alternative later.

**Comment by kzc** [comment]
Can `husky` and `eslint` be optional dependencies (or not installed) which can be occasionally run manually? It's only really needed before each release.

**Comment by fabiosantoscode** [comment]
I think there's no problem with using eslint (unless you try to commit while using node 0.10). Regarding husky, there are alternatives to it, namely [pre-commit](http://npmjs.com/pre-commit) that I haven't tried yet.

**Comment by fabiosantoscode** [comment]
@kzc pre-commit turned out to be an ES5 package, so there are no errors when the CI installs it.

Even though you obviously can't run eslint with node 0.10 (eslint is written in ES6), "npm install" now works because pre-commit's postinstall script is ES5.

**Comment by kzc** [comment]
@fabiosantoscode For future multi-commit PRs, could you please consider hitting the [Squash and merge] button? That way it's much easier to follow what's going on in the `master` commit log.

**Comment by fabiosantoscode** [comment]
Got it! Will do

>

**Pull request merged**